smt.go

```diff
@@ -1,457 +1,425 @@
-// Package smt implements a Sparse Merkle tree.
 package smt
 
 import (
 	"bytes"
-	"errors"
 	"hash"
 )
 
-const (
-	right = 1
+var (
+	_ treeNode = (*innerNode)(nil)
+	_ treeNode = (*leafNode)(nil)
 )
 
-var defaultValue = []byte{}
-
-var errKeyAlreadyEmpty = errors.New("key already empty")
-
-// SparseMerkleTree is a Sparse Merkle tree.
-type SparseMerkleTree struct {
-	th            treeHasher
-	nodes, values MapStore
-	root          []byte
+type innerNode struct {
+	leftChild, rightChild treeNode
+	persisted             bool
+	// Cached hash digest
+	digest []byte
 }
 
-// NewSparseMerkleTree creates a new Sparse Merkle tree on an empty MapStore.
-func NewSparseMerkleTree(nodes, values MapStore, hasher hash.Hash, options ...Option) *SparseMerkleTree {
-	smt := SparseMerkleTree{
-		th:     *newTreeHasher(hasher),
-		nodes:  nodes,
-		values: values,
-	}
-
-	for _, option := range options {
-		option(&smt)
-	}
-
-	smt.SetRoot(smt.th.placeholder())
-
-	return &smt
+type leafNode struct {
+	path      []byte
+	valueHash []byte
+	persisted bool
+	// Cached hash digest
+	digest []byte
 }
 
-// ImportSparseMerkleTree imports a Sparse Merkle tree from a non-empty MapStore.
-func ImportSparseMerkleTree(nodes, values MapStore, hasher hash.Hash, root []byte) *SparseMerkleTree {
-	smt := SparseMerkleTree{
-		th:     *newTreeHasher(hasher),
-		nodes:  nodes,
-		values: values,
-		root:   root,
-	}
-	return &smt
+// represents uncached persisted node
+type lazyNode struct {
+	digest []byte
 }
 
-// Root gets the root of the tree.
-func (smt *SparseMerkleTree) Root() []byte {
-	return smt.root
+type treeNode interface {
+	Persisted() bool
+	cachedDigest() []byte
 }
 
-// SetRoot sets the root of the tree.
-func (smt *SparseMerkleTree) SetRoot(root []byte) {
-	smt.root = root
+type SMT struct {
+	BaseSMT
+	nodes     MapStore
+	savedRoot []byte
+	// Current state of tree
+	tree treeNode
+	// Hashes of persisted nodes deleted from tree
+	orphans [][]byte
```

Comment:
Also wondering if we can optimize by minimizing copies and allocations by having a starting capacity here, and basically any other slice we have to continually resize.

Reply:
True. The expected capacity would depend on the tree size, so we can track size too.